src/app/teacher-round.service.ts

```diff
@@ -70,4 +88,86 @@ export class TeacherRoundService {
       endTime: this.endTime
     });
   }
+
+  assignBees(currentRoundPath: RoundPath, bees?: BeeWithWeight[]): void {
+    // Get the list of students in the session
+    this.firebaseService.getStudentsInSession(currentRoundPath.sessionId).pipe(
+      take(1),
+    ).subscribe(studentList => {
+      // If the round has a preset list of bees, use those
+      if (bees) {
+        this.customAssign(studentList, bees, currentRoundPath);
+      } else {
+        this.defaultAssign(studentList, currentRoundPath);
+      }
+    });
+  }
+
+  /**
+   * Unmark the currently active round, so that it's no longer active. Stop
+   * talking to the timer service.
+   *
+   * TODO: As of this iteration, this function just re-uses a single demo
+   * round and re-populates it with data, rather than creating a new round.
+   * We also don't unmark the currently active round yet.
+   */
+  endRound(sessionId: string): void {
+    this.roundPath$.next(null);
+  }
+
+  /**
+   * When no list of bees to use in the round is provided, randomly assign a bee to each student
+   * Note: It does so by randomizing the list of bees and using the new order to assign a bee to each student
+   * @param studentList the list of students to be assigned
+   * @param path the current round's path
+   */
+  defaultAssign(studentList: SessionStudentData[], path: RoundPath) {
+    const shuffledBees = this.shuffleArray(Object.values(allBeeSpecies));
+
+    studentList.forEach((student, studentIndex) => {
+      const beeIndex = studentIndex % shuffledBees.length;
+
+      this.firebaseService.addStudentToRound(student.id, path,
+        {beeSpecies: shuffledBees[beeIndex].id});
+    });
+  }
+
+  /**
+   * Assigns bees when list of bees is provided.
+   * Right now only works on demoBees.
+   */
+  customAssign(studentList: SessionStudentData[], beeList: BeeWithWeight[], path: RoundPath) {
+    // Shuffle the list of students to be a random order
+    const shuffledStudents = this.shuffleArray(studentList);
+
+    // Assign the students to a bee species based on the weight
+    let currentStudent = 0;
+    // TODO: Change to beeList once a proper round template is created
+    beeList.forEach(bee => {
+      const numStudents = Math.floor(bee.weight * shuffledStudents.length);
+      for (let i = currentStudent; i < currentStudent + numStudents; i++) {
+        this.firebaseService.addStudentToRound(shuffledStudents[i].id, path,
+          {beeSpecies: bee.id});
+      }
+      currentStudent += numStudents;
+    });
+
+    // Randomly assign the leftover students to a bee species
+    for (let i = currentStudent; i < shuffledStudents.length; i++) {
+      const beeIndex = Math.floor(Math.random() * demoBees.length);
+      this.firebaseService.addStudentToRound(shuffledStudents[i].id, path,
+        {beeSpecies: beeList[beeIndex].id});
+    }
+  }
+
+  shuffleArray(array: SessionStudentData[]): SessionStudentData[] {
```

Comment:
```suggestion
  shuffleArray<T>(array: T[]): T[] {
```

This would be a nifty place to use generics, by the way! :smiley: (That way, you can pass in both `SessionStudentData[]` either `BeeSpecies[]`, depending on what you want.)